### ISSUE TITLE
//g loops infinitely on tainted data

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#4855]
Migrated from [rt.perl.org#8262](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=8262) (status was 'resolved')

Searchable as RT8262$

**Comment by p5pRT** [comment]
### From skud@infotrope.net

#### Created by skud@infotrope.net



This is freaky\.  I've found an odd situation where Perl will go into an
infinite loop when doing a regexp match on tainted data\.  Here's the
simplest case I can put together&#8203;:

&nbsp;    \#\!/usr/bin/perl \-w

&nbsp;    use strict;

&nbsp;    print "The string we're using is '$ENV\{LANG\}'\\n";

&nbsp;    while\($ENV\{LANG\} =~ m/\(\.\)/g \) \{
&nbsp;        print "I'm stuck in a loop and I see letter '$1'\\n";
&nbsp;    \}


Try running that with taint checking off\, and you should see it work its
way through the letters in the language you're using\.  Then try it with 
tainting on\, and it'll get into an infinite loop\.

Cool\, huh?

There's no mention of this behaviour in perlsec\.pod\, so I'm figuring
it's a bug\.

For what it's worth\, this caught me in a Real World application where I
was using a CPAN module \(Locale&#8203;::Maketext\, by Sean Burke\) which
exhibited this unusual behaviour when I turned taint checking on\.

I've verified this behaviour with perl 5\.6\.0\, 5\.6\.1 and bleadperl\, all
on Debian\, and others have demonstrated it on other platforms\.

K\.


\`

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.7.2:

Configured by skud at Wed Jan 16 17:07:06 EST 2002.

Summary of my perl5 (revision 5.0 version 7 subversion 2 patch 14304) configuration:
  Platform:
    osname=linux, osvers=2.4.17, archname=i686-linux
    uname='linux tanqueray 2.4.17 #2 smp sun dec 23 14:01:48 est 2001 i686 unknown '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=define
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O3',
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.95.4 20011006 (Debian prerelease)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldbm -ldb -ldl -lm -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lcrypt -lutil
    libc=/lib/libc-2.2.4.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    DEVEL14289


@INC for perl v5.7.2:
    /home/skud/bleadperl-install/lib/5.7.2/i686-linux
    /home/skud/bleadperl-install/lib/5.7.2
    /home/skud/bleadperl-install/lib/site_perl/5.7.2/i686-linux
    /home/skud/bleadperl-install/lib/site_perl/5.7.2
    /home/skud/bleadperl-install/lib/site_perl
    /usr/local/share/perl/5.6.1/
    tanqueray
    ~> ls -l /usr/local/lib/perl5/site_perl/5.6.1/
    /usr/local/lib/perl5/site_perl/5.6.0//i686-linux
    /usr/local/lib/perl5/site_perl/5.6.0/
    .


Environment for perl v5.7.2:
    HOME=/home/skud
    LANG=en
    LANGUAGE=en
    LC_ALL=en_US
    LC_CTYPE=
    LD_LIBRARY_PATH=/usr/local/lib:/home/skud/rvplayer5.0
    LOGDIR (unset)
    PATH=/usr/local/jre1.3.1_01/bin:/usr/bin/mh:/usr/lib/mh:/usr/local/netscape:/usr/local/bin:/home/skud/rvplayer5.0:/usr/local/applix:/usr/X11R6/bin:/home/skud/bin:/usr/games:/usr/sbin:/sbin:/usr/local/jre1.3.1_01/bin:/usr/bin/mh:/usr/lib/mh:/usr/local/netscape:/usr/local/bin:/home/skud/rvplayer5.0:/usr/local/applix:/usr/X11R6/bin:/home/skud/bin:/usr/games:/usr/sbin:/sbin:/usr/local/jre1.3.1_01/bin:/usr/bin/mh:/usr/lib/mh:/usr/local/netscape:/usr/local/bin:/home/skud/rvplayer5.0:/usr/local/applix:/usr/X11R6/bin:/home/skud/bin:/usr/games:/usr/sbin:/sbin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games
    PERL_BADLANG (unset)
    SHELL=/usr/bin/tcsh


```
</details>

**Comment by p5pRT** [comment]
### From @vanstyn

Kirrily Robert \<skud@&#8203;infotrope\.net> wrote&#8203;:
\[paraphrased\]
&nbsp;  perl \-Twle 'print "\<$1>" while $ENV\{LANG\} =~ /\(\.\)/g'
loops forever\.

This seems to be a symptom of our current approach to localising magic\.
The point that we diverge between the with\-taint/no\-taint variants of
the above is when we get the helem for matching\, at the end of
pp\_hot\.c&#8203;:pp\_helem&#8203;:

&nbsp;    /\* This makes C\<local $tied\{foo\} = $tied\{foo\}> possible\.
&nbsp;     \* Pushing the magical RHS on to the stack is useless\, since
&nbsp;     \* that magic is soon destined to be misled by the local\(\)\,
&nbsp;     \* and thus the later pp\_sassign\(\) will fail to mg\_get\(\) the
&nbsp;     \* old value\.  This should also cure problems with delayed
&nbsp;     \* mg\_get\(\)s\.  GSAR 98\-07\-03 \*/
&nbsp;    if \(\!lval && SvGMAGICAL\(sv\)\)
&nbsp;        sv = sv\_mortalcopy\(sv\);

Because the taint adds a GMG flag\, we end up with the mortalcopy to
match against; we attach the pos\(\) magic to that\, and next time around
the loop we have a fresh mortalcopy and no pos\(\)\.

I don't know how to solve this\, but I suspect the solution is not
small\.

Hugo

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Kirrily Robert wrote&#8203;:
\[snip\]
>     while\($ENV\{LANG\} =~ m/\(\.\)/g \) \{
>         print "I'm stuck in a loop and I see letter '$1'\\n";
>     \}
> 
> Try running that with taint checking off\, and you should see it work
> its way through the letters in the language you're using\.  Then try it
> with tainting on\, and it'll get into an infinite loop\.
> 
> Cool\, huh?
> 
> There's no mention of this behaviour in perlsec\.pod\, so I'm figuring
> it's a bug\.

It is and it isn't a bug :\)

With tainting turned on\, $ENV\{LANG\} is slightly more magical than with
tainting turned off\, and each time you fetch its value\, it produces a
new scalar\.\.\. one with pos set to 0\.  As a workaround\, do&#8203;:
&nbsp;   my $x = $ENV\{LANG\};
&nbsp;   while\( $x =~ m/\(\.\)/g \) \{ \.\.\.\.

**Comment by p5pRT** [comment]
### From cmoore@hellyeah.org

This is also present in 5\.8\.0\.

\[skud \- Wed Jan 16 13&#8203;:03&#8203;:13 2002\]&#8203;:

> This is a bug report for perl from skud \<\!\-\-c\-\-> \<i>at\</i> \<\!\-\-a\-\-> infotrope\.net\,
> generated with the help of perlbug 1\.33 running under perl v5\.7\.2\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> This is freaky\.  I've found an odd situation where Perl will go into
>    an
> infinite loop when doing a regexp match on tainted data\.  Here's the
> simplest case I can put together&#8203;:
> 
>     \#\!/usr/bin/perl \-w
> 
>     use strict;
> 
>     print "The string we're using is '$ENV\{LANG\}'\\n";
> 
>     while\($ENV\{LANG\} =~ m/\(\.\)/g \) \{
>         print "I'm stuck in a loop and I see letter '$1'\\n";
>     \}
> 
> 
> Try running that with taint checking off\, and you should see it work
>    its
> way through the letters in the language you're using\.  Then try it
>    with
> tainting on\, and it'll get into an infinite loop\.
> 
> Cool\, huh?
> 
> There's no mention of this behaviour in perlsec\.pod\, so I'm figuring
> it's a bug\.
> 
> For what it's worth\, this caught me in a Real World application where
>    I
> was using a CPAN module \(Locale&#8203;::Maketext\, by Sean Burke\) which
> exhibited this unusual behaviour when I turned taint checking on\.
> 
> I've verified this behaviour with perl 5\.6\.0\, 5\.6\.1 and bleadperl\, all
> on Debian\, and others have demonstrated it on other platforms\.
> 
> K\.
> 
> 
> \`
> 
> 
> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\-
> Flags&#8203;:
>     category=core
>     severity=medium
> \-\-\-
> Site configuration information for perl v5\.7\.2&#8203;:
> 
> Configured by skud at Wed Jan 16 17&#8203;:07&#8203;:06 EST 2002\.
> 
> Summary of my perl5 \(revision 5\.0 version 7 subversion 2 patch 14304\)
>    configuration&#8203;:
>   Platform&#8203;:
>     osname=linux\, osvers=2\.4\.17\, archname=i686\-linux
>     uname='linux tanqueray 2\.4\.17 \#2 smp sun dec 23 14&#8203;:01&#8203;:48 est 2001
>    i686 unknown '
>     config\_args=''
>     hint=recommended\, useposix=true\, d\_sigaction=define
>     usethreads=undef use5005threads=undef useithreads=undef
>    usemultiplicity=undef
>     useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
>     use64bitint=undef use64bitall=undef uselongdouble=undef
>     usemymalloc=n\, bincompat5005=define
>   Compiler&#8203;:
>     cc='cc'\, ccflags ='\-fno\-strict\-aliasing \-I/usr/local/include
>    \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
>     optimize='\-O3'\,
>     cppflags='\-fno\-strict\-aliasing \-I/usr/local/include'
>     ccversion=''\, gccversion='2\.95\.4 20011006 \(Debian prerelease\)'\,
>    gccosandvers=''
>     intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
>     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\,
>    longdblsize=12
>     ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\,
>    lseeksize=8
>     alignbytes=4\, prototype=define
>   Linker and Libraries&#8203;:
>     ld='cc'\, ldflags =' \-L/usr/local/lib'
>     libpth=/usr/local/lib /lib /usr/lib
>     libs=\-lnsl \-lgdbm \-ldbm \-ldb \-ldl \-lm \-lc \-lcrypt \-lutil
>     perllibs=\-lnsl \-ldl \-lm \-lc \-lcrypt \-lutil
>     libc=/lib/libc\-2\.2\.4\.so\, so=so\, useshrplib=false\,
>    libperl=libperl\.a
>   Dynamic Linking&#8203;:
>     dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-
>    rdynamic'
>     cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'
> 
> Locally applied patches&#8203;:
>     DEVEL14289
> 
> \-\-\-
> @&#8203;INC for perl v5\.7\.2&#8203;:
>     /home/skud/bleadperl\-install/lib/5\.7\.2/i686\-linux
>     /home/skud/bleadperl\-install/lib/5\.7\.2
>     /home/skud/bleadperl\-install/lib/site\_perl/5\.7\.2/i686\-linux
>     /home/skud/bleadperl\-install/lib/site\_perl/5\.7\.2
>     /home/skud/bleadperl\-install/lib/site\_perl
>     /usr/local/share/perl/5\.6\.1/
>     tanqueray
>     ~> ls \-l /usr/local/lib/perl5/site\_perl/5\.6\.1/
>     /usr/local/lib/perl5/site\_perl/5\.6\.0//i686\-linux
>     /usr/local/lib/perl5/site\_perl/5\.6\.0/
>     \.
> 
> \-\-\-
> Environment for perl v5\.7\.2&#8203;:
>     HOME=/home/skud
>     LANG=en
>     LANGUAGE=en
>     LC\_ALL=en\_US
>     LC\_CTYPE=
>     LD\_LIBRARY\_PATH=/usr/local/lib&#8203;:/home/skud/rvplayer5\.0
>     LOGDIR \(unset\)
>     PATH=/usr/local/jre1\.3\.1\_01/bin&#8203;:/usr/bin/mh&#8203;:/usr/lib/mh&#8203;:/usr/local/netscape&#8203;:/
usr/local/bin&#8203;:/home/skud/rvplayer5\.0&#8203;:/usr/local/applix&#8203;:/usr/X11R6/bin&#8203;:/home/skud/
bin&#8203;:/usr/games&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/local/jre1\.3\.1\_01/bin&#8203;:/usr/bin/mh&#8203;:/usr/lib/mh&#8203;:/
usr/local/netscape&#8203;:/usr/local/bin&#8203;:/home/skud/rvplayer5\.0&#8203;:/usr/local/applix&#8203;:/usr/
X11R6/bin&#8203;:/home/skud/bin&#8203;:/usr/games&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/local/jre1\.3\.1\_01/bin&#8203;:/
usr/bin/mh&#8203;:/usr/lib/mh&#8203;:/usr/local/netscape&#8203;:/usr/local/bin&#8203;:/home/skud/rvplayer5\.0&#8203;:/
usr/local/applix&#8203;:/usr/X11R6/bin&#8203;:/home/skud/bin&#8203;:/usr/games&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/
local/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/bin/X11&#8203;:/usr/games
>     PERL\_BADLANG \(unset\)
>     SHELL=/usr/bin/tcsh
>

**Comment by p5pRT** [comment]
### From @smpeters

> \[skud \- Wed Jan 16 13&#8203;:03&#8203;:13 2002\]&#8203;:
> 
> This is a bug report for perl from skud@&#8203;infotrope\.net\,
> generated with the help of perlbug 1\.33 running under perl v5\.7\.2\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> This is freaky\.  I've found an odd situation where Perl will go into
>    an
> infinite loop when doing a regexp match on tainted data\.  Here's the
> simplest case I can put together&#8203;:
> 
>     \#\!/usr/bin/perl \-w
> 
>     use strict;
> 
>     print "The string we're using is '$ENV\{LANG\}'\\n";
> 
>     while\($ENV\{LANG\} =~ m/\(\.\)/g \) \{
>         print "I'm stuck in a loop and I see letter '$1'\\n";
>     \}
> 
> 
> Try running that with taint checking off\, and you should see it work
>    its
> way through the letters in the language you're using\.  Then try it
>    with
> tainting on\, and it'll get into an infinite loop\.
> 
> Cool\, huh?
> 
> There's no mention of this behaviour in perlsec\.pod\, so I'm figuring
> it's a bug\.
> 
> For what it's worth\, this caught me in a Real World application where
>    I
> was using a CPAN module \(Locale&#8203;::Maketext\, by Sean Burke\) which
> exhibited this unusual behaviour when I turned taint checking on\.
> 
> I've verified this behaviour with perl 5\.6\.0\, 5\.6\.1 and bleadperl\, all
> on Debian\, and others have demonstrated it on other platforms\.
> 
> K\.
> 
> 

With Perl\-5\.8\.7\, I see\.\.\.

steve@&#8203;kirk&#8203;:\~/perl\-current$ perl rt\_8262\.pl
The string we're using is 'en\_US\.UTF\-8'
I'm stuck in a loop and I see letter 'e'
I'm stuck in a loop and I see letter 'n'
I'm stuck in a loop and I see letter '\_'
I'm stuck in a loop and I see letter 'U'
I'm stuck in a loop and I see letter 'S'
I'm stuck in a loop and I see letter '\.'
I'm stuck in a loop and I see letter 'U'
I'm stuck in a loop and I see letter 'T'
I'm stuck in a loop and I see letter 'F'
I'm stuck in a loop and I see letter '\-'
I'm stuck in a loop and I see letter '8'
steve@&#8203;kirk&#8203;:\~/perl\-current$

Checking a few other recent Perls\, it looks like it went away somewhere
prior to 5\.8\.6\.

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From laocoon@fastmail.fm

> \[stmpeters \- Sat Dec 17 13&#8203;:23&#8203;:38 2005\]&#8203;:
> 
> > \[skud \- Wed Jan 16 13&#8203;:03&#8203;:13 2002\]&#8203;:
> >     \#\!/usr/bin/perl \-w
> > 
> >     use strict;
> > 
> >     print "The string we're using is '$ENV\{LANG\}'\\n";
> > 
> >     while\($ENV\{LANG\} =~ m/\(\.\)/g \) \{
> >         print "I'm stuck in a loop and I see letter '$1'\\n";
> >     \}
> > 
> > 
> > Try running that with taint checking off\, and you should see it work
> >    its
> > way through the letters in the language you're using\.  Then try it
> >    with
> > tainting on\, and it'll get into an infinite loop\.
> > 
> > Cool\, huh?
> > 
> > There's no mention of this behaviour in perlsec\.pod\, so I'm figuring
> > it's a bug\.
> > 
> 
> With Perl\-5\.8\.7\, I see\.\.\.
> 
> steve@&#8203;kirk&#8203;:\~/perl\-current$ perl rt\_8262\.pl
> The string we're using is 'en\_US\.UTF\-8'
> I'm stuck in a loop and I see letter 'e'
> I'm stuck in a loop and I see letter 'n'
> I'm stuck in a loop and I see letter '\_'
> I'm stuck in a loop and I see letter 'U'
> I'm stuck in a loop and I see letter 'S'
> I'm stuck in a loop and I see letter '\.'
> I'm stuck in a loop and I see letter 'U'
> I'm stuck in a loop and I see letter 'T'
> I'm stuck in a loop and I see letter 'F'
> I'm stuck in a loop and I see letter '\-'
> I'm stuck in a loop and I see letter '8'
> steve@&#8203;kirk&#8203;:\~/perl\-current$
> 
> Checking a few other recent Perls\, it looks like it went away somewhere
> prior to 5\.8\.6\.
>

Please note that this only occurs when running under \-T\.
Still loops here on 5\.8\.7 and blead\.

**Comment by p5pRT** [comment]
### From laocoon@fastmail.fm

> \[stmpeters \- Sat Dec 17 13&#8203;:23&#8203;:38 2005\]&#8203;:
> 
> > \[skud \- Wed Jan 16 13&#8203;:03&#8203;:13 2002\]&#8203;:
> >     \#\!/usr/bin/perl \-w
> > 
> >     use strict;
> > 
> >     print "The string we're using is '$ENV\{LANG\}'\\n";
> > 
> >     while\($ENV\{LANG\} =~ m/\(\.\)/g \) \{
> >         print "I'm stuck in a loop and I see letter '$1'\\n";
> >     \}
> > 
> > 
> > Try running that with taint checking off\, and you should see it work
> >    its
> > way through the letters in the language you're using\.  Then try it
> >    with
> > tainting on\, and it'll get into an infinite loop\.
> > 
> > Cool\, huh?
> > 
> > There's no mention of this behaviour in perlsec\.pod\, so I'm figuring
> > it's a bug\.
> > 
> 
> With Perl\-5\.8\.7\, I see\.\.\.
> 
> steve@&#8203;kirk&#8203;:\~/perl\-current$ perl rt\_8262\.pl
> The string we're using is 'en\_US\.UTF\-8'
> I'm stuck in a loop and I see letter 'e'
> I'm stuck in a loop and I see letter 'n'
> I'm stuck in a loop and I see letter '\_'
> I'm stuck in a loop and I see letter 'U'
> I'm stuck in a loop and I see letter 'S'
> I'm stuck in a loop and I see letter '\.'
> I'm stuck in a loop and I see letter 'U'
> I'm stuck in a loop and I see letter 'T'
> I'm stuck in a loop and I see letter 'F'
> I'm stuck in a loop and I see letter '\-'
> I'm stuck in a loop and I see letter '8'
> steve@&#8203;kirk&#8203;:\~/perl\-current$
> 
> Checking a few other recent Perls\, it looks like it went away somewhere
> prior to 5\.8\.6\.
>

Please note that this only occurs when running under \-T\.
Still loops here on 5\.8\.7 and blead\.

**Comment by p5pRT** [comment]
### From @ysth

On Sat\, Dec 17\, 2005 at 01&#8203;:23&#8203;:38PM \-0800\, Steve Peters via RT wrote&#8203;:
> > Try running that with taint checking off\, and you should see it work its
> > way through the letters in the language you're using\.  Then try it with
> > tainting on\, and it'll get into an infinite loop\.

I think you missed this part\, Steve\.

> steve@&#8203;kirk&#8203;:\~/perl\-current$ perl rt\_8262\.pl

&nbsp;                                  ^ \-T

> The string we're using is 'en\_US\.UTF\-8'
> I'm stuck in a loop and I see letter 'e'
> I'm stuck in a loop and I see letter 'n'
> I'm stuck in a loop and I see letter '\_'
> I'm stuck in a loop and I see letter 'U'
> I'm stuck in a loop and I see letter 'S'
> I'm stuck in a loop and I see letter '\.'
> I'm stuck in a loop and I see letter 'U'
> I'm stuck in a loop and I see letter 'T'
> I'm stuck in a loop and I see letter 'F'
> I'm stuck in a loop and I see letter '\-'
> I'm stuck in a loop and I see letter '8'
> steve@&#8203;kirk&#8203;:\~/perl\-current$
> 
> Checking a few other recent Perls\, it looks like it went away somewhere
> prior to 5\.8\.6\.

I see it failing in 5\.8\.7 and bleadperl \(\*with\* the \-T switch\)\.

**Comment by p5pRT** [comment]
### From BQW10602@nifty.com


On Sun\, 18 Dec 2005 02&#8203;:02&#8203;:34 \-0800\, Yitzchak Scott\-Thoennes \<sthoenna@&#8203;efn\.org> wrote

> On Sat\, Dec 17\, 2005 at 01&#8203;:23&#8203;:38PM \-0800\, Steve Peters via RT wrote&#8203;:
> > > Try running that with taint checking off\, and you should see it work its
> > > way through the letters in the language you're using\.  Then try it with
> > > tainting on\, and it'll get into an infinite loop\.
> 
> I think you missed this part\, Steve\.
> 
> > steve@&#8203;kirk&#8203;:\~/perl\-current$ perl rt\_8262\.pl
> 
>                                   ^ \-T

> I see it failing in 5\.8\.7 and bleadperl \(\*with\* the \-T switch\)\.

Yes\, the problem still remains to me\, too\.
But this problem seems to concern only a certain type of tainted values\,
but not all tainted values\.

When a value of %ENV is tainted\, its copy \(for example\, $a = $ENV\{LANG\}\)
is also tainted; but the copy doesn't cause the infinite loop\.

Similarly $ARGV\[0\] is harmful\, while its copy \($a = shift\) is harmless\.
A copy of \<> \(for example\, $a = \<STDIN>\) is harmless\.

Regards\,
SADAHIRO Tomoyuki

**Comment by p5pRT** [comment]
### From @petdance

>
> I see it failing in 5\.8\.7 and bleadperl \(\*with\* the \-T switch\)\.

You can also do it on&#8203;:

while \( $ENV\{LANG\} =~ m/\./g \) \{
&nbsp;     print "I'm stuck in the first loop and I see letter '$&'\\n";
\}


\-\-
Andy Lester => andy@&#8203;petdance\.com => www\.petdance\.com => AIM&#8203;:petdance

**Comment by p5pRT** [comment]
### From @smpeters

On Sun\, Dec 18\, 2005 at 11&#8203;:11&#8203;:05PM \-0800\, Andy Lester wrote&#8203;:
> >
> >I see it failing in 5\.8\.7 and bleadperl \(\*with\* the \-T switch\)\.
> 
> You can also do it on&#8203;:
> 
> while \( $ENV\{LANG\} =~ m/\./g \) \{
>     print "I'm stuck in the first loop and I see letter '$&'\\n";
> \}
> 

Yes\, I did miss the \-T\.  Reopened\.

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'resolved' to 'open'

**Comment by p5pRT** [comment]
### From mjtg@cam.ac.uk

SADAHIRO Tomoyuki \<bqw10602@&#8203;nifty\.com> wrote
> When a value of %ENV is tainted\, its copy \(for example\, $a = $ENV\{LANG\}\)
> is also tainted; but the copy doesn't cause the infinite loop\.

> Similarly $ARGV\[0\] is harmful\, while its copy \($a = shift\) is harmless\.
> A copy of \<> \(for example\, $a = \<STDIN>\) is harmless

So it looks like it's magic which causes the trouble \.\.\. presumably
magic\+taint is causing pos\(\) to be reset\.


Mike Guy

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Dec 19\, 2005 at 05&#8203;:41&#8203;:14PM \+0000\, Mike Guy wrote&#8203;:
> SADAHIRO Tomoyuki \<bqw10602@&#8203;nifty\.com> wrote
> > When a value of %ENV is tainted\, its copy \(for example\, $a = $ENV\{LANG\}\)
> > is also tainted; but the copy doesn't cause the infinite loop\.
> 
> > Similarly $ARGV\[0\] is harmful\, while its copy \($a = shift\) is harmless\.
> > A copy of \<> \(for example\, $a = \<STDIN>\) is harmless
> 
> So it looks like it's magic which causes the trouble \.\.\. presumably
> magic\+taint is causing pos\(\) to be reset\.

In something like

&nbsp;    $foo\[0\] =~ /bar/g

the expression on the LHS isn't treated as an lvalue\, and it so happens
that various ops \(such as aelem\) take liberties in this case and return a
mortal \*copy\* of the element if the element happens to have get magic\.

This is a problem with //g\, since it wants to modify the LHS by attaching
pos magic to it\, buts its axtually attaching the magic to the mortal copy\.

The patch below makes the LHS of a //g expression an lvalue\, and seems
to fix the problem\.

Dave

\-\- 
You live and learn \(although usually you just live\)\.

Change 26410 by davem@&#8203;davem\-splatty on 2005/12/19 22&#8203;:07&#8203;:49

&nbsp;    \[perl \#8262\] //g loops infinitely on tainted data
&nbsp;    make the LHS of expr =~ /foo/g an lvalue\, so that any pos magic
&nbsp;    attached to it stays attached\.

Affected files \.\.\.

\.\.\. //depot/perl/op\.c\#732 edit
\.\.\. //depot/perl/t/op/taint\.t\#69 edit

Differences \.\.\.

==== //depot/perl/op\.c\#732 \(text\) ====

@&#8203;@&#8203; \-1142\,8 \+1142\,9 @&#8203;@&#8203;
&nbsp;     /\* FALL THROUGH \*/
&nbsp;     default&#8203;:
&nbsp;       nomod&#8203;:
\-    /\* grep\, foreach\, subcalls\, refgen \*/
\-    if \(type == OP\_GREPSTART || type == OP\_ENTERSUB || type == OP\_REFGEN\)
\+    /\* grep\, foreach\, subcalls\, refgen\, m//g \*/
\+    if \(type == OP\_GREPSTART || type == OP\_ENTERSUB || type == OP\_REFGEN
\+        || type == OP\_MATCH\)
&nbsp;         break;
&nbsp;     yyerror\(Perl\_form\(aTHX\_ "Can't modify %s in %s"\,
&nbsp;              \(o\->op\_type == OP\_NULL && \(o\->op\_flags & OPf\_SPECIAL\)
@&#8203;@&#8203; \-1822\,9 \+1823\,14 @&#8203;@&#8203;
&nbsp;     \}
&nbsp;     if \(\!\(right\->op\_flags & OPf\_STACKED\) && ismatchop\) \{
&nbsp;     right\->op\_flags |= OPf\_STACKED;
\-    if \(right\->op\_type \!= OP\_MATCH &&
\-            \! \(right\->op\_type == OP\_TRANS &&
\-               right\->op\_private & OPpTRANS\_IDENTICAL\)\)
\+    /\* s/// and tr/// modify their arg\.
\+     \* m//g also indirectly modifies the arg by setting pos magic on it \*/
\+    if \(   \(right\->op\_type == OP\_MATCH &&
\+            \(cPMOPx\(right\)\->op\_pmflags & PMf\_GLOBAL\)\)
\+        || \(right\->op\_type == OP\_SUBST\)
\+        || \(right\->op\_type == OP\_TRANS &&
\+        \! \(right\->op\_private & OPpTRANS\_IDENTICAL\)\)
\+    \)
&nbsp;         left = mod\(left\, right\->op\_type\);
&nbsp;     if \(right\->op\_type == OP\_TRANS\)
&nbsp;         o = newBINOP\(OP\_NULL\, OPf\_STACKED\, scalar\(left\)\, right\);

==== //depot/perl/t/op/taint\.t\#69 \(xtext\) ====

@&#8203;@&#8203; \-17\,7 \+17\,7 @&#8203;@&#8203;
 use File&#8203;::Spec&#8203;::Functions;
 
 BEGIN \{ require '\./test\.pl'; \}
\-plan tests => 244;
\+plan tests => 245;
 
 
 $| = 1;
@&#8203;@&#8203; \-1135\,3 \+1135\,17 @&#8203;@&#8203;
&nbsp;     eval \{ local $0\, eval '1' \};
&nbsp;     test $@&#8203; eq '';
 \}
\+
\+\# \[perl \#8262\] //g loops infinitely on tainted data
\+
\+\{
\+    my @&#8203;a;
\+    $a\[0\] = $^X;
\+    my $i = 0;
\+    while\($a\[0\]=~ m/\(\.\)/g \) \{
\+    last if $i\+\+ > 10000;
\+    \}
\+    test $i \< 10000\, "infinite m//g";
\+    
\+\}
\+

**Comment by p5pRT** [comment]
### From @iabyn

Now fixed by commit fd69380d5d5b95ef16e2521cf4251b34ee0ce151
in branch davem/post\-5\.12\, which should be merged back into blead
once 5\.12 has been released\, and thus appear in 5\.13 onwards\.

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
